### PR TITLE
[TEX-537] Refactor HTTP request and response representations

### PR DESCRIPTION
**Pull request by timdawborn**
This PR does a bunch of initial refactoring and cleaning up of the representations of HTTP request and response objects throughout the codebase, ahead of making some larger changes to support gRPC-web request comparisons.

The changes in this PR are mostly code structure in nature rather than functional. There is one functionality change in the decision on how to compare two HTTP request bodies, which I've called out in comments.

### CODE REVIEW

File: src/similarity.ts [hunk 1]
```diff
@@ -13,75 +22,152 @@ import { Headers, PersistedBuffer, TapeRecord } from "./tape";
...
+      request1,
+      contentType1,
+      request2,
+      contentType2,
+      rewriteBeforeDiffRules,
+    );
+  } else {
+    // No more special cases to consider. Assume binary data for all other content types.
+    return countBodyDifferencesBinary(request1, request2);
+  }
```
Comment (timdawborn) [inline src/similarity.ts:110]: This is some slightly different logic to before.

Up until now, we just blindly attempted to decode it as JSON. If that succeeded, we did a JS object comparison of the resulting JSON decoded values. If it failed to decode, we did a base64 string comparison.

Now, we inspect the content type of the HTTP requests, and attempt to do JSON decoding if they have `application/json` as their content type, do raw string comparison if they have `text/<anything>` as their content type, and base64 comparison for any other content type.

File: src/similarity.ts [hunk 2]
```diff
@@ -120,19 +206,19 @@ function countStringDifferences(
...
     return path;
   }
 }
 
-function queryParameters(path: string) {
+function parseQueryParameters(path: string): ParsedUrlQuery {
   const questionMarkPosition = path.indexOf("?");
   if (questionMarkPosition !== -1) {
-    return queryString.parse(path.substr(questionMarkPosition));
+    return parseQueryString(path.substring(questionMarkPosition + 1));
```
Comment (timdawborn) [inline src/similarity.ts:221]: The ` + 1` addition here fixes a long standing bug that I found while swapping the query string parsing functionality from the long deprecated library to the builtin NodeJS library. Before, the substring returned included in the `?` at the start of the string as the slice was starting at the index of the `?`. Apparently the old library accepted this without issue, but the new NodeJS one did not.

File: package.json [hunk 3]
```diff
@@ -54,13 +54,14 @@
     "brotli": "^1.3.3",
     "chalk": "^4.1.0",
     "commander": "^10.0.1",
+    "content-type": "^1.0.5",
```
Comment (timdawborn) [inline package.json:57]: Added a HTTP `content-type` parsing library.

File: src/http.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import brotli from "brotli";
+import zlib from "zlib";
+import { ParsedMediaType as ParsedContentType } from "content-type";
+
+/**
+ * Headers of a request or response.
+ */
+export interface HttpHeaders {
+  [headerName: string]: string | string[] | undefined;
+}
```
Comment (timdawborn) [inline src/http.ts:10]: This has been moved from `tape.ts`.

File: src/http.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  host?: string;
+  method: string;
+  path: string;
+  headers: HttpHeaders;
+  body: Buffer;
+}
+
+export interface HttpRequestWithHost extends HttpRequest {
+  host: string;
+}
```
Comment (timdawborn) [inline src/http.ts:25]: These have been moved from `sender.ts`.

File: src/http.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * The common fields of a HTTP response.
+ */
+export interface HttpResponse {
+  status: {
+    code: number;
+  };
+  headers: HttpHeaders;
+  body: Buffer;
+}
```
Comment (timdawborn) [inline src/http.ts:36]: This has been moved from `tape.ts`.

**Review by njday — APPROVED** [review]
LGTM 👍

**Pull request merged**